diff -Naur ode-0.5.1/drawstuff/src/drawstuff.cpp ode-0.5_lpzrobots/drawstuff/src/drawstuff.cpp
--- ode-0.5.1/drawstuff/src/drawstuff.cpp	2005-09-13 15:27:20.000000000 +0200
+++ ode-0.5_lpzrobots/drawstuff/src/drawstuff.cpp	2005-09-19 16:01:44.000000000 +0200
@@ -74,11 +74,16 @@
 #define GROUND_G (0.5f)
 #define GROUND_B (0.3f)
 
+int groundTexture = DS_GROUND;
+int skyTexture = DS_SKY;
+
 const float ground_scale = 1.0f/1.0f;	// ground texture scale (1/size)
 const float ground_ofsx = 0.5;		// offset of ground texture
 const float ground_ofsy = 0.5;
 const float sky_scale = 1.0f/4.0f;	// sky texture scale (1/size)
 const float sky_height = 1.0f;		// sky height above viewpoint
+float sky_offset = 2.0f;                // offset of the sky texture. If 2.0 then use internal offset counter
+
 
 
 //***************************************************************************
@@ -880,7 +885,7 @@
   glDisable (GL_LIGHTING);
   if (use_textures) {
     glEnable (GL_TEXTURE_2D);
-    texture[DS_SKY]->bind (0);
+    texture[skyTexture]->bind (0);
   }
   else {
     glDisable (GL_TEXTURE_2D);
@@ -899,6 +904,15 @@
   float x = ssize*sky_scale;
   float z = view_xyz[2] + sky_height;
 
+  GLfloat fogColor[4] = {0.5, 0.5, 0.5, 1};
+  glEnable (GL_FOG);
+  glFogi (GL_FOG_MODE, GL_EXP2);
+  glFogfv (GL_FOG_COLOR, fogColor);
+  glFogf (GL_FOG_DENSITY, 0.02f);
+  glHint (GL_FOG_HINT, GL_NICEST); // GL_DONT_CARE);
+  glFogf (GL_FOG_START, 1.0);
+  glFogf (GL_FOG_END, 5.0);
+
   glBegin (GL_QUADS);
   glNormal3f (0,0,-1);
   glTexCoord2f (-x+offset,-x+offset);
@@ -911,8 +925,14 @@
   glVertex3f (ssize+view_xyz[0],-ssize+view_xyz[1],z);
   glEnd();
 
-  offset = offset + 0.002f;
-  if (offset > 1) offset -= 1;
+  if(sky_offset == 2.0f){
+    offset = offset + 0.002f;
+    if (offset > 1) offset -= 1;
+  }else{
+    offset=sky_offset;
+  }
+
+  glDisable (GL_FOG);
 
   glDepthFunc (GL_LESS);
   glDepthRange (0,1);
@@ -929,7 +949,7 @@
 
   if (use_textures) {
     glEnable (GL_TEXTURE_2D);
-    texture[DS_GROUND]->bind (0);
+    texture[groundTexture]->bind (0);
   }
   else {
     glDisable (GL_TEXTURE_2D);
@@ -937,18 +957,18 @@
   }
 
   // ground fog seems to cause problems with TNT2 under windows
-  /*
+  
   GLfloat fogColor[4] = {0.5, 0.5, 0.5, 1};
   glEnable (GL_FOG);
   glFogi (GL_FOG_MODE, GL_EXP2);
   glFogfv (GL_FOG_COLOR, fogColor);
-  glFogf (GL_FOG_DENSITY, 0.05f);
+  glFogf (GL_FOG_DENSITY, 0.02f);
   glHint (GL_FOG_HINT, GL_NICEST); // GL_DONT_CARE);
   glFogf (GL_FOG_START, 1.0);
   glFogf (GL_FOG_END, 5.0);
-  */
+  
 
-  const float gsize = 100.0f;
+  const float gsize = 300.0f;
   const float offset = 0; // -0.001f; ... polygon offsetting doesn't work well
 
   glBegin (GL_QUADS);
@@ -1187,7 +1207,7 @@
   glDisable (GL_LIGHTING);
   if (use_textures) {
     glEnable (GL_TEXTURE_2D);
-    texture[DS_GROUND]->bind(1);
+    texture[groundTexture]->bind(1);
     glColor3f (SHADOW_INTENSITY,SHADOW_INTENSITY,SHADOW_INTENSITY);    
     glEnable (GL_TEXTURE_2D);
     if(genTextureCoords){
@@ -1283,6 +1303,17 @@
   }
 }
 
+extern "C" void dsSetGroundTexture(int _texture){
+  if(_texture >0 && _texture < DS_MAXTEXTURES && texture[_texture])
+    groundTexture = _texture;
+}
+
+extern "C" void dsSetSkyTexture(int _texture){
+  if(_texture >0 && _texture < DS_MAXTEXTURES && texture[_texture])
+    skyTexture = _texture;
+}
+
+
 extern "C" int dsRegisterTexture (const char* filename, bool uselocalpath)
 {
   if (!texture_prefix) dsError ("No texture prefix defined");
@@ -1306,6 +1337,11 @@
   return (texture_number-1);
 }
 
+/* set the sky_offset */
+extern "C" void dsSetSimulationTime(double time){
+  double t = time/10.0;
+  sky_offset=(t - long(t))*2.0-1.0; // sky_offset is between -1 and 1. (this is basically time modulo 10 scaled to -1:1)
+}
 
 extern "C" void dsSetColor (float red, float green, float blue)
 {
@@ -1367,7 +1403,7 @@
   if (use_shadows) {
     glDisable (GL_LIGHTING);
     if (use_textures) {
-      texture[DS_GROUND]->bind(1);
+      texture[groundTexture]->bind(1);
       glEnable (GL_TEXTURE_2D);
       glDisable (GL_TEXTURE_GEN_S);
       glDisable (GL_TEXTURE_GEN_T);
diff -Naur ode-0.5.1/include/drawstuff/drawstuff.h ode-0.5_lpzrobots/include/drawstuff/drawstuff.h
--- ode-0.5.1/include/drawstuff/drawstuff.h	2005-09-13 15:27:20.000000000 +0200
+++ ode-0.5_lpzrobots/include/drawstuff/drawstuff.h	2005-09-19 16:23:25.000000000 +0200
@@ -58,7 +58,6 @@
 #define DS_WOOD   3
 #define DS_MAXTEXTURES 128
 
-
 typedef struct dsFunctions {
   int version;			/* put DS_VERSION here */
   /* version 1 data */
@@ -105,6 +104,7 @@
  */
 void dsStop();
 
+
 /* Registers a user defined texture and returns its number. This can be used 
  * as parameter to dsSetTexture()   
  * @param uselocalpath use local path or add prefix texture path
@@ -122,6 +122,15 @@
  */
 void dsSetTexture (int texture_number, int _coating = 1);
 
+/* set the textures for ground and sky. texture is a value returned by dsRegisterTexture*/ 
+void dsSetGroundTexture(int texture);
+void dsSetSkyTexture(int texture);
+
+/* sets the offset for the sky image. This makes it possible to let the cloud speed to reflect the true time 
+   If you don't use it then the offset modified in each draw step
+*/  
+void dsSetSimulationTime(double time);
+
   /* if side = 0 then BackFace else FrontFace*/
 void dsSetCulling(int side);
 void dsSetColor (float red, float green, float blue);
